lib/Index/IndexRecord.cpp

```diff
@@ -495,6 +495,22 @@ emitDataForSwiftSerializedModule(ModuleDecl *module,
                                  IndexUnitWriter &parentUnitWriter,
                                  const PathRemapper &pathRemapper,
                                  SourceFile *initialFile) {
+  // Reload resilient modules from swiftinterface to avoid indexing
+  // internal details.
+  if (module->getResilienceStrategy() == ResilienceStrategy::Resilient) {
+    module->getASTContext().setIgnoreAdjacentModules(true);
+
+    ImportPath::Module::Builder builder(module->getName());
+    auto reloadedModule = module->getASTContext().getModule(builder.get());
```

Comment:
I'm concerned about the cost of re-reading the module here.  Even if nothing is later deserialized from the module, my understanding is that loading the module has a non-trivial cost (this was important in code-completion for example).  Indexing-while-building is enabled by default in all debug builds, so this would have a wide impact.

Can we move the reload of the module below the `parentUnitWriter.isUnitUpToDateForOutputFile` check below? Then we would only incur this cost when we will actually need to index the given module. We would still need the updated module filename for that check, but not to actually read the file.

Reply:
Good point, I moved the new logic after the "Up to date check". I also tweaked it in the hope the we still write down something when the swiftinterface fails to build, so the next indexing phase won't attempt to rebuild it again.

I'm not sure how to test superfluous reindexing though, please let me know if you have an idea.

Reply:
> I'm not sure how to test superfluous reindexing though, please let me know if you have an idea.

We could add a remark (`indexing-system-module`), add `-Rindexing-system-module` to the test, and then check the output of the second compile doesn't contain it?

Reply:
Good idea, I added this remark and used it in the test. It shows on attempts to index, whenever we write a file, and specifies if the content is fixed because of a broken swiftinterface.